include/mbedtls/rsa.h

```diff
@@ -583,21 +562,39 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
                          unsigned int nbits, int exponent );
 
 /**
- * \brief          Check if a context contains an RSA public key
+ * \brief          Check if a context contains (at least) an RSA public key
  *
  * \param ctx      RSA context to be checked
  *
- * \return         0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code
+ * \return         0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code.
+ *                 On success, it is guaranteed that enough information is
+ *                 present to perform an RSA public key operation
+ *                 \c mbedtls_rsa_public.
+ *
  */
 int mbedtls_rsa_check_pubkey( const mbedtls_rsa_context *ctx );
 
 /**
- * \brief          Check if a context contains a complete
- *                 and valid RSA private key.
+ * \brief          Check if a context contains an RSA private key
+ *                 and perform basic sanity checks.
  *
  * \param ctx      RSA context to be checked
  *
- * \return         0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code
+ * \return         0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code.
+ *                 On success, it is guaranteed that enough information is
+ *                 present to perform RSA private and public key operations.
+ *
+ * \warning        This function is *not* obliged to perform an exhaustive
+ *                 sanity check what would guarantee the internal parameters
+ *                 to match and \c mbedtls_rsa_private and \c mbedtls_rsa_public
+ *                 to be mutually inverse to each other.
+ *                 The reason is that for minimal non-CRT implementations
+ *                 using only N, D, E, for example, checking the validity
+ *                 would be computationally expensive.
```

Comment:
@gilles-peskine-arm With the current implementation, there's a strong difference between what the documentation mandates and what the implementation actually does. The implementation doesn't guarantee anything, as you and @mpg suggested, because it simplifies life when it comes to reduced contexts. The current implementation, though, which still has all the fields around for ABI compatibility, performs a full consistency check as it was before. 

My preference would be to strengthen the documentation at least a little bit to enforce something more than just a "the RSA functions won't crash" like a heuristic check whether private and public key match, like a sample encryption-decryption. Such a heuristic check could be performed even for reduced contexts. But I found it hard converting this into a halfway precise requirement, so I sticked to the very weak version as it is now, while the implementation still performs the full check.

Reply:
@hanno-arm My preference is the opposite: don't try to do anything more than you're promising. I don't expect this validation function to include any heuristics, all I want is a hard guarantee that `mbedtls_rsa_xxx` functions won't choke if I call them on this key.

Reply:
I don't feel comfortable with not providing the user any means to have more confidence in the private key material than knowing that the system won't crash when using it. While trusting the key material goes beyond what's effectively verifiable, I'd nonetheless like to have a function which performs some functional correctness check. 

@gilles-peskine-arm @yanesca, what do you think of the following approach: The `mbedtls_rsa-complete` function, which any user must call anyway, guarantees to have performed the 'crash-test' checks when it succeeds, so that the user can continue to use the RSA primitives knowing that at least they won't crash. In contrast, `mbedtls_rsa_check_privkey` performs a stronger key-check, which might be an empirical check consisting of an example decryption-encryption as in the PK alt interface, or - as we're doing now - a validity check of the RSA parameters in case they are all present. 

With that separation, the user may choose if he wants to also perform the `check_privkey`, and moreover, the semantics of the latter stays closer to what it has been before: Even if the documentation is [so blurry](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/rsa.h#L180) that it would be allowed, I'd prefer to not switch from a function which currently still performs a full key-check to one which does only trivial divide-by-zero like checks.

Reply:
@hanno-arm Yes, I think having only the no-crash check in the function that callers must call (`mbedtls_rsa_complete`), and providing another function that isn't part of the basic workflow to perform other checks. `pkparse` should only perform the no-crash check.

Reply:
@gilles-peskine-arm, @yanesca Suggestion for a documentation of `rsa_complete`:
```
 * \return
 *                 - 0 if successful. In this case, it is guaranteed
 *                   that the RSA context can be used for RSA operations
 *                   without the risk of failure or crash.
 *                 - \c MBEDTLS_ERR_RSA_BAD_INPUT_DATA if the attempted
 *                   derivations failed.
 *
 * \warning        Implementations are *not* obliged to perform
 *                 consistency checks for the imported parameters!
 *                 In particular, parameters that are not needed by the
 *                 implementation may be silently discarded and left
 *                 unchecked. For the purpose of checking the consistency
 *                 of the key material, see \c mbedtls_rsa_check_privkey.
```
Suggestion for a documentation of `rsa_check_privkey`:
```
 * \note       This function performs checks substantiating
 *             the consistency of the key material used to setup
 *             the RSA context. In case of implementations saving
 *             all core RSA parameters, this might mean a consistency
 *             check in the sense of \c mbedtls_rsa_validate_params,
 *             while other implementations might perform an empirical
 *             check consisting of an encryption-decryption pair.
 *
 * \warning    This function should catch accidental misconfigurations
 *             like swapping of parameters, but it cannot establish full
 *             trust in neither the quality nor the consistency of the key
 *             material that was used to setup the given RSA context:
 *             - Regarding consistency, note (see \c mbedtls_rsa_complete)
 *               that imported parameters irrelevant for the implementation
 *               might be silently dropped, in which case the present
 *               function doesn't have access to and hence cannot check them.
 *               If the user desires to check the consistency of the entire
 *               content of, say, an PKCS1-encoded RSA private key, he
 *               should use \c mbedtls_rsa_validate_params before setting
 *               up the RSA context.
 *               Further, if the implementation performs empirical checks,
 *               these checks will substantiate but not guarantee consistency.
 *             - Regarding quality, this function is not expected to perform
 *               extended quality assessments like checking that the prime
 *               factors are safe. Further, it is the user's responsibility to
 *               ensure trustworthiness of the source of his RSA parameters,
 *               a question going beyond what's effectively checkable
 *               by the library.
```

Reply:
@hanno-arm I like the approach you agreed on with @gilles-peskine-arm.

Reply:
I agree with Gilles about performing only no-crash checks in the basic workflow, and providing a separate function for more extended checks.

I'm no sure the wording of the documentation is optimal as its stand, eg "Implementations are not obliged too" seems to be mainly targeted at implementers of alternative implementations and might be confusing to someone who just wants to use the library. I think it's great (and overdue) that we're now providing clearer guidelines to alt-implementers but this should not come at the expense of potentially confusing simple library users.